### ISSUE TITLE
 [Errno 104] Connection reset by peer.

### ISSUE THREAD

**Issue by vedavidhbudimuri** [erlio/vernemq#441]
### Environment

While load testing the facing [Errno 104] Connection reset by peer. I have spawned 100 processes with each creating 100 threading and each thread sending 100 messages. I'm using a single node here on a 16 GB system with 4 cores.

I'm using erlang 20, vernemq 1.1.0, OS Ubuntu

Here is the console.log 
2017-07-15 10:53:00.349 [error] <0.238.0>@vmq_ranch_config:reconfigure_listeners_for_type:187 can't reconfigure mqtt listener({127,0,0,1}, 1883) with Options [{max_connections,infinity},{nr_of_acceptors,200000},{mountpoint,[]},{proxy_protocol,false}] due to {{shutdown,{failed_to_start_child,ranch_acceptors_sup,{listen_error,{{127,0,0,1},1883},eaddrinuse}}},{child,undefined,{ranch_listener_sup,{{127,0,0,1},1883}},{ranch_listener_sup,start_link,[{{127,0,0,1},1883},200000,ranch_tcp,[{ip,{127,0,0,1}},{port,1883},{nodelay,true},{linger,{true,0}},{send_timeout,30000},{send_timeout_close,true}],vmq_ranch,[{mountpoint,[]}]]},permanent,infinity,supervisor,[ranch_listener_sup]}}
2017-07-15 10:53:00.349 [error] <0.342.0> Supervisor {<0.342.0>,ranch_listener_sup} had child ranch_acceptors_sup started with ranch_acceptors_sup:start_link({{127,0,0,1},1883}, 200000, ranch_tcp, [{ip,{127,0,0,1}},{port,1883},{nodelay,true},{linger,{true,0}},{send_timeout,30000},{send_timeout_close,...}]) at undefined exit with reason {listen_error,{{127,0,0,1},1883},eaddrinuse} in context start_error

I'm unable to use vernemq to its full potentials as in various issues its mentioned that server hundreds of thousands of connections simultaneously, But here im facing issue with just 10k connections.

**Comment by ioolkos** [comment]
Hi @vedavidhbudimuri thanks for your report... it seems that VerneMQ cannot setup the default MQTT listener here, as your 1883 port on localhost is already in use.
(I take that you want to run a test from localhost clients to localhost broker)

**Comment by vedavidhbudimuri** [comment]
Yes im testing from localhost. What could be the fix? for it

**Comment by ioolkos** [comment]
Well, some process is running owning port 1883. (possibly another MQTT broker...possibly another VerneMQ process). You need to find and terminate that process before starting VerneMQ (or change to another port in vernemq.conf, but I suggest you do the first solution).

**Comment by vedavidhbudimuri** [comment]
Yeah i have made sure that there isn't any other process owning that port. But this time I have encountered the below crash. With Connection [Errno 104] Connection reset by peer Error at client 

```
2017-07-15 11:42:09 =CRASH REPORT====
  crasher:
    initial call: vmq_cluster_mon:init/1
    pid: <0.389.0>
    registered_name: vmq_cluster_mon
    exception exit: {noproc,[{gen_server,terminate,7,[{file,"gen_server.erl"},{line,800}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [vmq_cluster_node_sup,vmq_server_sup,<0.238.0>]
    messages: []
    links: [<0.356.0>]
    dictionary: []
    trap_exit: true
    status: running
    heap_size: 376
    stack_size: 27
    reductions: 276
  neighbours:
2017-07-15 11:42:09 =SUPERVISOR REPORT====
     Supervisor: {local,vmq_cluster_node_sup}
     Context:    child_terminated
     Reason:     noproc
     Offender:   [{pid,<0.389.0>},{id,vmq_cluster_mon},{mfargs,{vmq_cluster_mon,start_link,[]}},{restart_type,permanent},{shutdown,5000},{child_type,worker}]

2017-07-15 11:42:09 =ERROR REPORT====
** Generic server vmq_cluster_mon terminating
** Last message in was timeout
** When Server state == {state}
** Reason for termination ==
** {noproc,[{gen,do_for_proc,2,[{file,"gen.erl"},{line,261}]},{gen_event,rpc,2,[{file,"gen_event.erl"},{line,197}]},{vmq_cluster_mon,terminate,2,[{file,"/opt/vernemq/distdir/1.1.0/_build/default/lib/vmq_server/src/vmq_cluster_mon.erl"},{line,153}]},{gen_server,try_terminate,3,[{file,"gen_server.erl"},{line,629}]},{gen_server,terminate,7,[{file,"gen_server.erl"},{line,795}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
```

**Comment by vedavidhbudimuri** [comment]
Duplicate of #442

**Comment by ioolkos** [comment]
Is this a throw away test deployment? If yes, I'd delete all the content of `/var/lib/vernemq/meta` and restart the broker. Just to make sure to test for a clean state. Your previous testing might actually have overloaded the broker with unintended consequences for the restart. 
I'm just guessing as I don't know how your are testing.

**Comment by vedavidhbudimuri** [comment]
Yeah I have tried just the way you said. The logs seems to be clean but facing  [Errno 104] Connection reset by peer. Im just trying to publish and subscribe to the same topic. And each thread  has its own unique topic.

**Comment by ioolkos** [comment]
Note on the vernemq.conf:
- Start with lower max_online and max_offline messages. Setting it to 200k means per session! do a quick calculation what this might mean for your RAM.
- Do not configure 200k acceptors. You should be good with 1000. Rule of thumb should be that you don't need more than your expected connection setup rate per second.

As far as your error goes: what client lib, what testing tool are you using? is it something I can reproduce?

**Comment by vedavidhbudimuri** [comment]
Yeah Im using paho python client. The script is simple just created 100 processes with each creating  100 threads and each publishing 100 messages without any sleep function calls.

**Comment by ioolkos** [comment]
Make sure you use different ClientIDs for every connection in your python script. (you probably do, just making sure).
Also: start with a much lower test scenario (1 connection for example). What happens? seeing the same error? then it's probably related to some client lib problem/misunderstanding.

**Comment by vedavidhbudimuri** [comment]
Here is the python script that im trying to run on my server. 
I have attached the script as txt file. To execute it you need to install this package before executing the script
https://pypi.python.org/pypi/paho-mqtt/1.1

[my_script.txt](https://github.com/erlio/vernemq/files/1150346/my_script.txt)

**Comment by ioolkos** [comment]
Thanks, only had a quick look and failed with `NameError: global name 'new_messages_dict' is not defined`. Probably harmless but didn't try further.

Other than that, in general, try a controlled test setup rate, that is not hitting the broker with everything at once.

**Comment by vedavidhbudimuri** [comment]
My bad I have updated the script which is fine. Yeah its now hitting the broker at a time.

**Comment by larshesel** [comment]
Hi @vedavidhbudimuri it looks like from the other issues you opened that you were able to connect successfully to VerneMQ. As the current discussion doesn't have anything to do with the original issue I'll close this (and thus remove it from the list of open issues). Feel free to continue the discussion, though.

**Comment by vedavidhbudimuri** [comment]
Hey @larshesel ,
I'm facing the issue while having 10k connections, in the other issues i have just created only one connection just for testing. So can you help me with this issue?

**Comment by larshesel** [comment]
We've never seen the Errno 104 connection reset problem before. what happens exactly when you see this problem? Does VerneMQ itself crash completely, are all clients disconnected or are only new clients trying to connect rejected?

Can you show us the contents of the console.log file after you've experienced this issue? (please delete the log files before starting VerneMQ and doing the experiment so there is no old data in them).

**Comment by vedavidhbudimuri** [comment]
Hey @larshesel ,
Only a few clients that are trying to connect are rejected (not all the clients that are trying to connect). I guess it's working fine with a load of 5k connections.

**Comment by larshesel** [comment]
My guess is you're running into some kind of system limit which prevents you from creating more connections. But without more concrete information it's hard to say more. To answer your question from https://github.com/erlio/vernemq/issues/441#issuecomment-315547068: VerneMQ has been tested to be able several hundred thousands of connections on a single node - so you should not have any issues to connect more than 10k clients.

**Comment by vedavidhbudimuri** [comment]
My system configuration is 8GB RAM with 4 cores.

**Comment by larshesel** [comment]
Please provide the console.log file as described above.

**Comment by vedavidhbudimuri** [comment]
Hey @larshesel ,
There are two cases, You can find the logs, configs file and the script im using to test in each of the scenario zip files.
[scenario_1.zip](https://github.com/erlio/vernemq/files/1152167/scenario_1.zip)
[scenario_2.zip](https://github.com/erlio/vernemq/files/1152174/scenario_2.zip)

**Comment by larshesel** [comment]
In the first you do not have proper authorization rules in place and the publishes are not allowed (is `vmq_acl` properly configured). In the second you've run into a system limit: 

    2017-07-17 09:50:03.513 [error] <0.352.0> can't load acl file "/etc/vernemq/vmq.acl" due to emfile

the `emfile` error means that it was not possible open the file because the system limit on the number of open files was hit. See http://vernemq.com/docs/misc/change-open-file-limits.html#change-open-files-limit

**Comment by vedavidhbudimuri** [comment]
Yes, @larshesel I too noticed that and followed the same doc. You could see the limits in the screenshot I have put in the scenario_2.zip, which is more than whats mentioned in the documentation. Isn't that fine? or do I have to set it to any other particular value?

**Comment by larshesel** [comment]
You have to make sure this limit applies to the vernemq process. If vernemq is running try to run `os:cmd("ulimit -n")` to check what limit the vernemq process is observing :

```
vernemq attach
Remote Shell: Use "Ctrl-G q" to quit.
q() or init:stop() will terminate the vernemq node.
Erlang/OTP 19 [erts-8.3.5] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V8.3.5  (abort with ^G)
(VerneMQ@127.0.0.1)1> os:cmd("ulimit -n").
"1024\n"
(VerneMQ@127.0.0.1)2> 
User switch command
 --> q
```

**Comment by vedavidhbudimuri** [comment]
Hey @larshesel followed the steps again but its still the same unable to set to vernemq process.

**Comment by vedavidhbudimuri** [comment]
Hey @larshesel 

What does this logs actually mean, I have encountered this after increasing the load to 100K and on the client side error: [Errno 98] Address already in use. Does this mean? 
`2017-07-17 11:26:55.212 [warning] <0.226.0>@plumtree_broadcast:schedule_lazy_tick:599 Broadcast overloaded, 23ms mailbox traversal, schedule next lazy broadcast in 22134ms, the min interval is 10000ms
2017-07-17 11:29:41.886 [warning] <0.226.0>@plumtree_broadcast:schedule_lazy_tick:599 Broadcast overloaded, 13ms mailbox traversal, schedule next lazy broadcast in 19343ms, the min interval is 10000ms
2017-07-17 11:30:37.297 [warning] <0.226.0>@plumtree_broadcast:schedule_lazy_tick:599 Broadcast overloaded, 27ms mailbox traversal, schedule next lazy broadcast in 23003ms, the min interval is 10000ms
2017-07-17 11:31:57.270 [warning] <0.226.0>@plumtree_broadcast:schedule_lazy_tick:599 Broadcast overloaded, 20ms mailbox traversal, schedule next lazy broadcast in 21407ms, the min interval is 10000ms
2017-07-17 11:32:00.706 [warning] <0.3252.15>@vmq_ranch:teardown:127 session stopped abnormally due to 'badarg'
2017-07-17 11:32:52.522 [warning] <0.226.0>@plumtree_broadcast:schedule_lazy_tick:599 Broadcast overloaded, 57ms mailbox traversal, schedule next lazy broadcast in 27597ms, the min interval is 10000ms
2017-07-17 11:32:57.371 [warning] <0.29230.15>@vmq_ranch:teardown:127 session stopped abnormally due to 'badarg'
2017-07-17 11:33:02.088 [warning] <0.17427.16>@vmq_ranch:teardown:127 session stopped abnormally due to 'badarg'
2017-07-17 11:33:02.091 [warning] <0.25971.14>@vmq_ranch:teardown:127 session stopped abnormally due to 'badarg'
2017-07-17 11:33:02.146 [warning] <0.16505.16>@vmq_ranch:teardown:127 session stopped abnormally due to 'badarg'
2017-07-17 11:33:18.561 [warning] <0.18742.17>@vmq_ranch:teardown:127 session stopped abnormally due to 'badarg'
2017-07-17 11:33:20.560 [warning] <0.21845.17>@vmq_ranch:teardown:127 session stopped abnormally due to 'badarg'
2017-07-17 11:34:23.958 [warning] <0.226.0>@plumtree_broadcast:schedule_lazy_tick:599 Broadcast overloaded, 48ms mailbox traversal, schedule next lazy broadcast in 26458ms, the min interval is 10000ms
`

**Comment by larshesel** [comment]
https://stackoverflow.com/questions/4465959/python-errno-98-address-already-in-use

It's not related to the above logs.

**Comment by vedavidhbudimuri** [comment]
@larshesel thanks for the reply.

**Comment by Kemwal1** [comment]
hello, everyone,
i have two scripts written on python (publisher / subscriber) i want to send a simple message on port 8883, i got this error message [Errno 104] reset by a peer.
I looked at the file consol.log to see the file, I could not understand here is the text file .. can someone help me ???

thank you in advance
[Publisher - Subscriber 1.txt](https://github.com/vernemq/vernemq/files/2855310/Publisher.-.Subscriber.1.txt)
[Consol.log.txt](https://github.com/vernemq/vernemq/files/2855309/Consol.log.txt)

**Comment by larshesel** [comment]
@Kemwal1 there are no errors in the vernemq log files. you could try to trace the client-id (see `vmq-admin trace --help`) to see what's going on on the server side - likely it's a configuration problem llke either a network issue or authentication/authorization issue. For more help our slack channel would likely be the best option,

**Comment by Kemwal1** [comment]
@larshesel thx for your quick reply.  
I want to know if the scripts are OK ?
